### PR TITLE
HLD for getting error status of the SPF modules

### PR DESCRIPTION
**Pull request by stephenxs**
Initial version

Signed-off-by: Stephen Sun <stephens@nvidia.com>

### CODE REVIEW

File: doc/mellanox/sfp-error-status.md [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+```
+admin@sonic:~$ show platform mlnx transceiver error-status --port Ethernet8
+Port       Error Status
+---------  ------------------------------------
+Ethernet8  No error
+
+admin@sonic:~$ show platform mlnx transceiver error-status 
+Port         Error Status
+-----------  ----------------------------------------------
+Ethernet0    No error
```
Comment (keboliu) [inline doc/mellanox/sfp-error-status.md]: "No error" -> "OK"

File: doc/mellanox/sfp-error-status.md [hunk 2]
```diff
@@ -0,0 +1,195 @@
+# Fetch SFP error status through CLI #
```
Comment (keboliu) [inline doc/mellanox/sfp-error-status.md]: Fetch SFP error status through CLI on Mellanox Platfom

File: doc/mellanox/sfp-error-status.md [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+## Table of Content
+
+### Revision
+
+### Scope
+
+This is the high level design for the feature "fetch SFP error status through CLI".
```
Comment (keboliu) [inline doc/mellanox/sfp-error-status.md]: fetch SFP error status through CLI on Mellanox platform

File: doc/mellanox/sfp-error-status.md [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+### Overview
+
+The purpose of this section is to give an overview of high-level design document and its architecture implementation in SONiC.
+
+### Requirements
+
+The requirement:
+
+- To expose the SFP error status to CLI
+- One shot command for debugging, which means the error status should be directly polled from low lever, eg. SDK, instead of being cached in the database.
```
Comment (keboliu) [inline doc/mellanox/sfp-error-status.md]: poll -> get/fetch/query?

File: doc/mellanox/sfp-error-status.md [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+- Only supported on Mellanox system. The command will be under `show platform mlnx`
+
+### Architecture Design
+
+CLI to call platform API directly and platform API to fetch error status by calling SDK API.
+
+- A new platform API is introduced to fetch the error code of the SPF module via using the SDK API.
```
Comment (keboliu) [inline doc/mellanox/sfp-error-status.md]: this API is vendor-specific, we should make it clear

File: doc/mellanox/sfp-error-status.md [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+- A CLI command is introduced to display the error status to the user. Since the platform API is available in pmon docker only, the CLI has to call platform API via `docker exec` and then parse the output of the platform API.
+- No database change is required.
+
+### High-Level Design
+
+#### Platform API
+
+A new platform API in class SFP is required for polling the error status of the SFP module.
```
Comment (keboliu) [inline doc/mellanox/sfp-error-status.md]: replace polling?

File: doc/mellanox/sfp-error-status.md [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+---------  ------------------------------------
+Ethernet8  No error
+
+admin@sonic:~$ show platform mlnx transceiver error-status 
+Port         Error Status
+-----------  ----------------------------------------------
+Ethernet0    No error
+Ethernet4    No error
+Ethernet8    No error
+Ethernet12   No error
```
Comment (keboliu) [inline doc/mellanox/sfp-error-status.md]: is it possible to have multi errors? can we give an example if yes?  can we give an example if SFP is not present?

Comment (stephenxs) [reply doc/mellanox/sfp-error-status.md]: Only one error is supported for each module.
If SFP is not present, the error code should be "OK" as well.

**Comment by thuffmlx** [comment]
OK, great stuff, but having a return status of 0 mean "POWER BUDGET EXCEEDED" and return code 255 mean "No error" breaks a very common convention of 0 return code means "No errors". Recommend that you change this.

**Comment by stephenxs** [comment]
> OK, great stuff, but having a return status of 0 mean "POWER BUDGET EXCEEDED" and return code 255 mean "No error" breaks a very common convention of 0 return code means "No errors". Recommend that you change this.

Thank you for your comments.
The error codes and error descriptions, like 0 for “power budget exceeded” and 255 for “reserved”, are taken from the SDK API (which takes from the PRM). 
As the feature is Mellanox only and this is an internal interface inside platform API, I don’t tend to change them for now.
But if we are going to make this kind of feature a common one and the error code and interface will be exposed to the users, I’ll take it into consideration.

**Comment by jleveque** [comment]
Is there a way we can move this logic to xcvrd? This way we can keep in line with the SONiC ecosystem (e.g., xcvrd is the only application responsible for communicating with transceivers), and it also opens up the ability for other vendors to implement this if they have similar support on their platforms.

**Comment by stephenxs** [comment]
> Is there a way we can move this logic to xcvrd? This way we can keep in line with the SONiC ecosystem (e.g., xcvrd is the only application responsible for communicating with transceivers), and it also opens up the ability for other vendors to implement this if they have similar support on their platforms.

I think there are two approaches to achieve it:
1. Periodically:
   - `xcvrd` to fetch the error status periodically and save the possible error status into `STATE_DB`
   - CLI to fetch error status from `STATE_DB` each time the command is called.
2. Synchronous communication:
   - A dedicated database table, probably in `STATE_DB`, is introduced for communication between the backend, `xcvrd`, and the frontend, the CLI, or other possible UI.
   - `xcvrd` listens to an update on the database table. On receiving the message, it fetches the error status and stores it into `STATE_DB`
   - CLI updates the database table, indicating it wants `xcvrd` to fetch the error code for some SFP modules and listens to the table for the reply. It will display the error status to the user once it receives the update from the table.
   - Proper timeout mechanism should be introduced as well.

As this is a command for debugging, it may have real-time requirements. So the second approach is preferred.

Meanwhile, I'm not sure whether this is the scope of this design or another design should be opened for this

**Pull request closed without merging**